core/app/services/spree/promotions/code_generator.rb

```diff
@@ -0,0 +1,80 @@
+module Spree
+  module Promotions
+    class CodeGenerator
+      MutuallyExclusiveInputsError = Class.new(StandardError)
+
+      def initialize(config = {})
+        @config = config
+      end
+
+      def build
+        validate_inputs if config[:deny_list]
+        loop do
+          candidate = compose
+          break candidate if valid?(candidate)
+        end
+      end
+
+      private
+
+      attr_reader :config
+
+      def validate_inputs
+        raise_error if inputs_invalid?
+      end
+
+      def valid?(subject)
+        return true if config[:deny_list].nil?
+
+        violation_checks = config[:deny_list].map do |el|
+          subject.include?(el)
+        end
+
+        !violation_checks.any?
+      end
+
+      def compose
+        case config[:affix]
+        when :prefix
+          prefix_alorithm
+        when :suffix
+          suffix_alorithm
+        else
+          default_algorithm
+        end
+      end
+
+      def prefix_alorithm
+        clear_result.concat(content, random_code)
+      end
+
+      def suffix_alorithm
+        clear_result.concat(random_code, content)
+      end
+
+      def default_algorithm
+        random_code
+      end
+
+      def content
+        config[:content]
+      end
+
+      def random_code
+        SecureRandom.hex(4)
+      end
+
+      def clear_result
+        result = ""
```

Comment:
👍 Good catch. Using `+` instead of `concat` renders this unnecessary. I've added it because `concat` mutates its base, so in this case it was e.g. appending the code, to the base altered in the previous step of the loop.